src/components/TodoApp/TodoApp.tsx

```diff
@@ -0,0 +1,141 @@
+import React, {
+  useCallback, useContext, useEffect,
+} from 'react';
+import cn from 'classnames';
+import { TodosContext } from '../../store/TodoProvider';
+import { TodosFilter } from '../TodosFilter/TodosFilter';
+import { TodoList } from '../TodoList/TodoList';
+import { TodoForm } from '../TodoForm/TodoForm';
+import {
+  ActionType, ErrorMessage, FilterType,
+} from '../../types/Todo';
+import { removeTodo } from '../../api/todos';
+
+export const TodoApp: React.FC = () => {
+  const { state, dispatch } = useContext(TodosContext);
+
+  const { todos, tempTodo, error } = state;
+
+  const toggleAllTodoItems = () => {
+    dispatch({ type: ActionType.TOGGLE_ALL });
+  };
+
+  const removeCompletedTodoItems = () => {
+    const completedTodosIds = todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    dispatch({
+      type: ActionType.PROCESSED,
+      payload: completedTodosIds,
+    });
+
+    completedTodosIds.forEach(id => {
+      removeTodo(id)
+        .then(() => dispatch({
+          type: ActionType.REMOVE,
+          payload: id,
+        }))
+        .catch(() => dispatch({
+          type: ActionType.ERROR,
+          payload: ErrorMessage.UNABLE_DELETE,
+        }));
+    });
+  };
+
+  const setFilterType = (filterBy: FilterType) => {
+    dispatch({ type: ActionType.FILTER, payload: filterBy });
+  };
+
+  const clearError = useCallback(() => {
+    dispatch({ type: ActionType.ERROR, payload: '' });
+  }, [dispatch]);
+
+  const filteredTodos = () => {
+    switch (state.filterBy) {
+      case FilterType.ACTIVE:
+        return todos.filter(item => !item.completed);
+      case FilterType.COMPLETED:
+        return todos.filter(item => item.completed);
+      default:
+        return todos;
+    }
+  };
+
+  useEffect(() => {
+    let timerId = 0;
+
+    if (error) {
+      timerId = window.setTimeout(() => {
+        clearError();
+      }, 3000);
+    }
+
+    return () => {
+      window.clearTimeout(timerId);
+    };
+  }, [error, clearError]);
+
+  return (
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          {/* eslint-disable jsx-a11y/control-has-associated-label */}
+          {todos.length > 0 && (
+            <button
+              type="button"
+              className={cn('todoapp__toggle-all', {
+                active: todos.every(item => item.completed),
+              })}
+              onClick={() => toggleAllTodoItems()}
+              data-cy="ToggleAllButton"
+            />
+          )}
+
+          <TodoForm />
+        </header>
+
+        <TodoList items={filteredTodos()} tempItem={tempTodo} />
```

Comment:
```suggestion
        <TodoList items={filteredTodos} tempItem={tempTodo} />
```
when you will wrap in useMemo filtered todos - no need will be call filteredTodos here (it will be array instead function)